src/QsCompiler/Tests.Compiler/TestCases/CapabilityVerification.qs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/// Test cases for verification of execution target runtime capabilities.
+namespace Microsoft.Quantum.Testing.CapabilityVerification {
+    internal operation X(q : Qubit) : Unit {
+        body intrinsic;
+    }
+
+    internal operation M(q : Qubit) : Result {
+        body intrinsic;
+    }
+
+    function ResultAsBool(result : Result) : Bool {
+        return result == Zero ? false | true;
+    }
+    
+    function ResultAsBoolNeq(result : Result) : Bool {
+        return result != One ? false | true;
+    }
+
+    operation ResultAsBoolOp(result : Result) : Bool {
+        return result == Zero ? false | true;
+    }
+    
+    function ResultAsBoolNeqOp(result : Result) : Bool {
+        return result != One ? false | true;
+    }
+
+    operation ResultAsBoolOpReturnIf(result : Result) : Bool {
+        if (result == Zero) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+    
+    operation ResultAsBoolNeqOpReturnIf(result : Result) : Bool {
+        if (result != One) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    operation ResultAsBoolOpSetIf(result : Result) : Bool {
+        mutable b = false;
+        if (result == One) {
+            set b = true;
+        }
```

Comment:
One quirk just to be aware of: somewhere in the compiler, the if-statements get transformed into an if-else on a comparison to zero. So 
```
if (result == One) {
    set b = true;
}
```
will effectively become:
```
if (result == Zero) {}
else {
    set b = true;
}
```
I don't think this invalidates this testing mechanism you've written, and in fact it makes sense to write the Q# code as the user would, since this transformation is a compiler detail and may change over time. Just calling it out so you can have that in the back of your mind for the future!

Reply:
There are operations for control flow like [`ApplyIfOne`](https://github.com/microsoft/qsharp-runtime/blob/master/src/Simulation/QsharpCore/ClassicalControl.qs#L130) that are used by the [classical control transformation](https://github.com/microsoft/qsharp-compiler/blob/master/src/QsCompiler/Transformations/ClassicallyControlled.cs#L408). I thought that is what the compiler would turn those into?

Reply:
It does that classical rewrite step only when an `ExecutionTarget` is specified, and when it does, those calls still seem to get inserted so that it is logically equivalent to an if-else with comparison against `Zero`. I'm still not sure exactly where that happens though, as I haven't taken the time to dig into that part of the compiler yet. That's what I meant by it will "effectively become" the other statement, since it doesn't literally get transformed into that Q#. The only reason I know this is from stepping through and being surprised that a sample with `let res == One; if (res == One) { <do something> }` ended up in an else clause handler.

Reply:
Having the different test cases is still useful here as these tests should be agnostic to the classical control conversions.

Reply:
OK, makes sense, thanks for the explanation. I agree that it's good to test all the cases since the classical control details can change.